### PR TITLE
feat: Adds separateMultipleMinor

### PR DESCRIPTION
**Pull request by ccopsey**
## Changes:

Adds separateMultipleMinor option, which creates a branch/PR for each available minor upgrade, for each existing major upgrade.  The modifications account for individual packages and groups, across minor and major upgrades.  It also only adds this feature, it does not alter the behaviour of existing functionality.

## Context:

#6576

Attempting to apply a stabilityDays critera to a package which is released frequently means that you only ever get a PR for the final patch version of the final minor version per major version.  This feature should allow the final patch version of each minor version to become an upgrade target.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

See https://github.com/ccopsey-renovate/separateMultipleMinor for an example of a repo that has been renovated with this feature.

### CODE REVIEW

File: lib/workers/repository/updates/branch-name.ts
Comment (rarkins) [inline lib/workers/repository/updates/branch-name.ts:46]: If you need to do this then it probably means there's a types problem elsewhere

**Review by rarkins — CHANGES_REQUESTED** [review]
What's the expected behavior when multiple major releases are available but separateMultipleMajor=false while separateMultipleMinor=true?

E.g. let's say you are on 1.0.0 and 2.0.0 and 3.0.0 are available. Today you'd just get one PR - for 3.0.0. If you set separateMultipleMinor would you still get just one PR our would you get an extra one for 2.0.0 too?

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2450]: This needs more explanation than just one hard to understand sentence.
Can you make the behavior visual, by making a list or table?

I'm thinking of something like this:

> If you enable `separateMultipleMinor` and you have a package `foo` with versions `x.y.z`, `x.y.z` and `x.y.z` then you'll get update PRs for these versions:
> 
> - List version 1
> - List version 2

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2448]: We have a section in the FAQ that explains how to "Separate patch releases from minor releases".

The new feature in this PR seems related to this concept as well?
Do we want a new section in the FAQ that links to the new `separateMultipleMinor` config documentation?

https://docs.renovatebot.com/faq/#separate-patch-releases-from-minor-releases

File: lib/util/template/index.ts
Comment (ccopsey) [reply lib/util/template/index.ts]: Hi, my intention was that this option should have no impact on which major versions should be candidates for upgrade. It should just allow minors within the already established set of majors to become candidates.

To answer your question, I would say 3.0.x, 3.1.x are candidates. 2.x.x would not. 

Following your question I tried running my updates with separateMultipleMajor=false, and the result can be seen here: (https://github.com/ccopsey-renovate/separateMultipleMinor2)

The output wasn’t exactly as I had hoped, as major/minor combos generated PRs where the major version shouldn’t have been a candidate but the minor >= minor. I haven’t had time to review yet.

The complex naming of branches and PRs depending on all the permutations of options means that anything more complicated than what I have attempted is probably out of my scope for the foreseeable future.

File: docs/usage/configuration-options.md
Comment (ccopsey) [reply docs/usage/configuration-options.md:2450]: I agree, however I didn’t want to put too much effort into the docs until the premise behind what I attempted was agreed.

Comment (HonkingGoose) [reply docs/usage/configuration-options.md:2450]: That's okay! I understand wanting to know if the idea is good or not before putting effort into the docs. 😄 

Please expand the documentation once you have approval for your code. 😉

**Comment by danports** [comment]
Hi folks, is there any way I could help get this PR over the finish line? I'd like to see it merged particularly for Kubernetes, where it's recommended that you not skip over any minor versions when upgrading.

**Comment by rarkins** [comment]
Let's go back to the requirements issue and make sure we agree on behavior in each scenario and then we can move this PR forward

**Comment by ccopsey** [comment]
When I started looking at this feature my intention was to keep the change as small as possible.  I'm not a Node/JS dev and didn't want to bite off too much.  And I believe this was broadly inline with the conclusion in #6576.  As I worked through the code it seemed that this whole concept was ripe for an overhaul, and I was not the guy for that.  Regardless, I was determined to finish what I started, which is what you see above.

With that in mind, my intent was to add a single new bool which allowed minor upgrades to become candidates in a way that was fully backward compatible.  So if a major version was a candidate, minor versions within that major version could become candidates.

In order to answer @rarkins question from 08/12 I prepped a new repo to demo what separateMultipleMajor = false, separateMultipleMinor = true would do.  Given a current version of 1.0.0, and available versions 2.0.0, 2.1.0, 3.0.0, 3.1.0 I would expect PRs for 3.0.0 and 3.1.0.  All the 2.x would be skipped as separateMultipleMajor didn't allow them.

[separateMultipleMinor2](https://github.com/ccopsey-renovate/separateMultipleMinor2)

Unfortunately the behaviour wasn't quite right.  Only minor versions with a value greater than the major version become PRs.  And before I knew it 2 months had past and here we are.

While preparing my changes and becoming confident with the code I ran a series of tests to determine the effect of all permutations of separateMultipleMajor, separateMajorMinor, and separateMinorPatch.  The results can be seen [here](https://github.com/ccopsey-renovate/separateDocs).

Overall - I'm still desperate for this feature.  I use renovate to upgrade renovate itself, with a 7 day stability, and I'm still on 30.4.1 from 17/12 simply because there hasn't been another major release in the interim.  I think the best approach would be for someone more capable than I to sort this out properly/thoroughly/confidently.  Failing that, I will get around to fixing this PR one day.  Not soon though unfortunately.

File: lib/workers/repository/process/lookup/bucket.ts
Comment (rarkins) [inline lib/workers/repository/process/lookup/bucket.ts:37]:
```suggestion
```

No changes if major updates

File: lib/workers/repository/process/lookup/bucket.ts
Comment (rarkins) [inline lib/workers/repository/process/lookup/bucket.ts:42]:
```suggestion
```

No changes if major updates

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
Closing due to conflicts and previous comment. We would still like to move this forward though.

**Pull request closed without merging**
